Improving helper scripts and example documentation for mysql/mariadb

hhorak opened the pull request:
Based on the feedback from OpenShift, these patches rename the commands available in the container to container-usage and container-entrypoint.

It is also shown how common documentation may be defined for all images, while every image may add own specific documentation.

The documentation is then available when running `container-usage` and also available in the README.md in every image.

`container-usage` is also set as `CMD` by default, but may be changed for every image.

hhorak committed:
Rename cont-entry to container-entrypoint and cont-help to container-usage since it is more self-explanating

hhorak committed:
Changes in the help/usage:
 - rename the file to container-usage
 - use -ti in case of docker exec CONTAINERID bash
 - specify container-usage for all images if not re-defined

hhorak committed:
Allow to have one common documentation for container-usage and for Dockerfile README.md

hhorak committed:
Add MySQL and MariaDB specific documentation

hhorak commented:
@mfojtik This is mysql/mariadb example of how the doc was meant to look. It also includes changes in command names we talked about yesterday.

hhorak committed:
Include change proposed by https://github.com/sclorg/rhscl2dockerfile/pull/3/commits

hhorak committed:
Interpret /usr/share/cont-docs/70-general.txt as template, since it is a template

praiskup commented:
Looks OK.

hhorak commented:
I agree to show example with container-usage every-time.

mfojtik commented on common/cont-lib/usr/share/cont-docs/70-general.txt.tpl:
```diff
@@ -0,0 +1,16 @@
+
+Run `docker run -ti {% if privileged %} --privileged {% endif -%} {{ container }} bash` to obtain interactive shell.
+
+Run `docker exec -ti {% if privileged %} --privileged {% endif -%} CONTAINERID bash` to access already running container.
```
why you need to pass the --privileged ?

praiskup replied:
Good point, thats probably not needed for the basic usage described
in "General container help".

hhorak replied:
just devtoolset-3-systemtap needs it, it was requested by @humaton maybe he knows more about the concrete use case.

mfojtik replied:
i won't go the way suggesting --privileged to all images just because one tool needs it ;-) will rather patch/fix that tool to tell what it needs.

praiskup replied:
@mfojtik, while talking about this particular diff, its just template - only
selected images would have '--privileged' at the end.

But yes, I can't think of image where --privileged would be needed for 'container-usage'.

hhorak replied:
Just to be clear, the information that the container must be run as privileged container is available only in the one image (devtoolset-3-systemtap) after the files are generated.
If you suggest to remove it from this file, then it needs to be included in the documentation specific for the image. We don't have such a documentation for devtoolset-3-systemtap yet, so we'd need to create one. Is that what we need?

praiskup replied:
@hhorak IMO yes.  Because the general info about the container describes ATM
three cases which should - even on *-systemtap images - be OK to run without
--privileged.

Or we can say something like:

```
{% if privileged %}
This container is expected to be run like:
    'docker run --privileged {{ container }}'
{% endif %}
```

mfojtik commented on common/cont-lib/usr/share/cont-docs/70-general.txt.tpl:
```diff
@@ -0,0 +1,16 @@
+info.
+
+
+Report bugs to <http://bugzilla.redhat.com>.
```
maybe include the email from MAINTAINER field as well

hhorak replied:
I don't think we want to get any reports to any email address, do we?

mfojtik commented:
@praiskup @hhorak will the usage be printed when I do 'docker run ruby_image' ? (without any cmd specified)

praiskup commented:
@mfojtik  wrote:

> @praiskup @hhorak will the usage be printed when I do 'docker run ruby_image'
> ? (without any cmd specified)

Yes.

hhorak commented:
@praiskup Including the `readme` file automatically without specifying  -- I like this idea but I'm not sure how to do it, since the generator doesn't know where the file is for specific image..

hhorak committed:
Advise to run container-usage every time

praiskup commented:
@hhorak me neither, probably some additional jinja filter for '*.txt' pattern inclusion in cont-docs would be needed.  Its not meant to be fixed in this pull request.

praiskup commented:
FYI, the cont-lib part: https://github.com/devexp-db/cont-lib/pull/1

hhorak committed:
Add special documentation for devtoolset-3-systemtap image, so we may get rid of the privileged condition

hhorak committed:
Remove jinja tags from 70-general.txt so it may be the same as in upstream cont-lib
We should not use specific image name in the help, since the image name is defined in a different place and users may re-tag it
And when user accesses help in the container, the name of the image/container is known already, thus using THIS_IMAGE

hhorak committed:
Revert removing container-entrypoint, since it is actually the only way scl will be available every-time

hhorak commented:
@praiskup @mfojtik I tried to update the pull request with all ideas to adjust (if not forgot something), are we ready to merge?

mfojtik commented on common/cont-lib/usr/share/cont-docs/70-general.txt:
```diff
@@ -0,0 +1,16 @@
+
+Run `docker run -ti THIS_IMAGE bash` to obtain interactive shell.
+
+Run `docker exec -ti CONTAINERID container-entrypoint` to access already running container.
```
we still don't like the `container-entrypoint` helper as enabler for SCL collections.... 

@smarterclayton do you know if there is something we can do in `docker exec` to set the SCL vars before you enter container namespace?

bparees replied:
+1

praiskup replied:
@mfojtik wrote:

> we still don't like the container-entrypoint helper as enabler for SCL
> collections....

Its not ideal, yes.  But doing

```
`docker exec CONTID container-entrypoint your command`
```

... instead of ...

```
`docker exec CONTID your command`
```

does not seem to be _that_ bad.  We could consider this to be a temporary
workaround (untill Docker allows us to do that better way).

Pavel

mfojtik commented on common/cont-lib/usr/share/cont-docs/70-general.txt:
```diff
@@ -0,0 +1,16 @@
+General container help
+----------------------
+
+Run `docker run THIS_IMAGE container-usage` to get this help.
```
It is possible to have two version of this help? One for platform/language images, where this helper is not needed and second for database images where you have to override the default CMD to get the help? This is not a strong requirement, we will most likely be doing it in OpenShift by layering this image with STI scripts and poiting the CMD to `usage` script.

praiskup replied:
@mfojtik wrote:

> It is possible to have two version of this help?  One for platform/language
> images, where this helper is not needed and second for database images where
> you have to override the default CMD to get the help?
> This is not a strong requirement, we will most likely be doing it in
> OpenShift by layering this image with STI scripts and poiting the CMD to
> usage script.

If you do `docker run THIS_IMAGE` on platform/language images, you'll get
the 'container-usage' automatically.  Do I understand your concerns right?

Those commands are equivalent on "platform/language" images:

```
  docker run THIS_IMAGE
  docker run THIS_IMAGE container-usage
```

But IMO its better to stay explicit in the help output.

Pavel

praiskup replied:
To not forget, if you add something like
/usr/share/cont-docs/50-openshift-httpd.txt in your additional layer, you'll
get your help text into 'container-usage' automatically.

mfojtik commented:
@hhorak LGTM, just last nit about the container-entrypoint for docker exec.

hhorak commented:
@mfojtik just to make it clear, `docker exec -ti CONT bash` will work as well, but user may simply get an impression that `docker exec -ti CONT /bin/sh` or `docker exec -ti CONT script.py` will work as well, which is not true (we cannot solve this particular case with SCL). Having `docker exec -ti CONT container-entrypoint` in the usage should make it more safe for user.
